### ISSUE TITLE
request hangs without timeout

### ISSUE THREAD

**Issue by cstjean** [JuliaWeb/HTTP.jl#417]
I have a query `HTTP.get(url; status_exception=false, readtimeout=5*60)` that hung for 25 minutes, no CPU. When I interrupted it with Ctrl-C, I got this stacktrace:

```julia
 [1] process_events at ./libuv.jl:98 [inlined]
 [2] wait() at ./event.jl:246
 [3] wait(::Condition) at ./event.jl:46
 [4] wait_readnb(::Sockets.TCPSocket, ::Int64) at ./stream.jl:368
 [5] eof at ./stream.jl:61 [inlined]
 [6] read_to_buffer(::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::Int64) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/ConnectionPool.jl:204
 [7] readuntil(::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::Function, ::Int64) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/ConnectionPool.jl:224
 [8] readuntil at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/ConnectionPool.jl:222 [inlined]
 [9] readheaders(::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::HTTP.Messages.Response) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/Messages.jl:473
 [10] startread(::HTTP.Streams.Stream{HTTP.Messages.Response,HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}}) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/Streams.jl:153
 [11] macro expansion at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/StreamRequest.jl:65 [inlined]
 [12] macro expansion at ./task.jl:244 [inlined]
 [13] #request#1(::Nothing, ::Nothing, ::Int64, ::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:status_exception, :readtimeout),Tuple{Bool,Int64}}}, ::Function, ::Type{HTTP.StreamRequest.StreamLayer}, ::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::HTTP.Messages.Request, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/StreamRequest.jl:54
 [14] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:iofunction, :status_exception, :readtimeout),Tuple{Nothing,Bool,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.StreamRequest.StreamLayer}, ::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::HTTP.Messages.Request, ::Array{UInt8,1}) at ./none:0
 [15] #request#1(::Nothing, ::Type, ::Int64, ::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:iofunction, :status_exception, :readtimeout),Tuple{Nothing,Bool,Int64}}}, ::Function, ::Type{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer}}, ::HTTP.URIs.URI, ::HTTP.Messages.Request, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/ConnectionRequest.jl:56
 [16] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:iofunction, :status_exception, :readtimeout),Tuple{Nothing,Bool,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer}}, ::HTTP.URIs.URI, ::HTTP.Messages.Request, ::Array{UInt8,1}) at ./none:0
 [17] (::getfield(Base, Symbol("###48#49#50")){ExponentialBackOff,getfield(HTTP.RetryRequest, Symbol("##2#3")){Bool,HTTP.Messages.Request},typeof(HTTP.request)})(::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:iofunction, :status_exception, :readtimeout),Tuple{Nothing,Bool,Int64}}}, ::Function, ::Type, ::Vararg{Any,N} where N) at ./error.jl:231
 [18] ##48#51 at ./none:0 [inlined]
 [19] #request#1 at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/RetryRequest.jl:44 [inlined]
 [20] #request at ./none:0 [inlined]
 [21] #request#1(::VersionNumber, ::String, ::Nothing, ::Nothing, ::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:status_exception, :readtimeout),Tuple{Bool,Int64}}}, ::Function, ::Type{HTTP.MessageRequest.MessageLayer{HTTP.RetryRequest.RetryLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/MessageRequest.jl:47
 [22] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:status_exception, :readtimeout),Tuple{Bool,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.MessageRequest.MessageLayer{HTTP.RetryRequest.RetryLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at ./none:0
 [23] #request#1(::Int64, ::Bool, ::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:status_exception, :readtimeout),Tuple{Bool,Int64}}}, ::Function, ::Type{HTTP.RedirectRequest.RedirectLayer{HTTP.MessageRequest.MessageLayer{HTTP.RetryRequest.RetryLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer}}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/RedirectRequest.jl:24
 [24] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:status_exception, :readtimeout),Tuple{Bool,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.RedirectRequest.RedirectLayer{HTTP.MessageRequest.MessageLayer{HTTP.RetryRequest.RetryLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.StreamRequest.StreamLayer}}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at ./none:0
 [25] #request#5(::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:status_exception, :readtimeout),Tuple{Bool,Int64}}}, ::Function, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/HTTP.jl:300
 [26] #request at ./none:0 [inlined]
 [27] #request#6 at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/HTTP.jl:314 [inlined]
 [28] #request at ./none:0 [inlined] (repeats 2 times)
 [29] #get#13 at /home/cst-jean/.julia/packages/HTTP/GN0Te/src/HTTP.jl:382 [inlined]
 [30] #get at ./none:0 [inlined]
```

On HTTP v0.8.0, Julia 1.1.1, Linux x86. Is that expected? It seems like the timeout is not respected. It's not exactly 100% reproducible, but I make a lot of similar requests, and I've had it hang 4 times already.

**Comment by quinnj** [comment]
So one thing I noticed looking into this is that the default retry limit for idempotent requests is 4, so in your case, I'd at least expect the total time to be ~20 minutes (5 minute readtimeout * 4 retries). Another thing I noticed is that the default sleep for retry is `8 + rand() * 4`, so readtimeout shorter than that won't even work. I'm not sure if we need to necessarily support that, but it was annoying because https://httpbin.org/delay/10 only supports a max delay of 10 seconds.

@cstjean, do you have more details on the specific request you can share? Is it a request I could reproduce locally (even if you must share details privately).

**Comment by cstjean** [comment]
We've boiled it down somewhat. We set up a rest service in C#, using the WCF library, and we put a debugger break point in the request response code. On the Julia side, 

```julia
julia> using HTTP

julia> HTTP.get(url, retry=false, retries=1, readtimeout=5)
```
hangs, for 30 minutes and counting. cc @fcalvez

Setting up a public server isn't trivial in our position, unfortunately.

**Comment by cstjean** [comment]
We made a small server that demonstrates the issue, https://github.com/cstjean/BlockingServer.js

```julia
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.1.1 (2019-05-16)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> versioninfo()
]stJulia Version 1.1.1
Commit 55e36cc308 (2019-05-16 04:10 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, haswell)
Environment:
  JULIA_EDITOR = emacsclient

(v1.1) pkg> st HTTP
    Status `~/.julia/environments/v1.1/Project.toml`
  [cd3eb016] HTTP v0.8.0

julia> using HTTP

julia> url = "http://letivaz:8082/nolock/"
"http://letivaz:8082/nolock/"

julia> HTTP.get(url, retry=false, retries=1, readtimeout=15)
HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
X-Powered-By: Express
Content-Type: application/json
Date: Wed, 05 Jun 2019 14:54:46 GMT
Connection: keep-alive
Content-Length: 17

{"value":"12345"}"""

julia> url = "http://localhost:8082/lock/120"
"http://localhost:8082/lock/120"

julia> HTTP.get(url, retry=false, retries=1, readtimeout=15)
HTTP.Messages.Response:
"""
HTTP/1.1 200 OK
X-Powered-By: Express
Content-Type: application/json
Date: Wed, 05 Jun 2019 14:57:24 GMT
Connection: keep-alive
Content-Length: 18

{"value":"locked"}"""
```
Why didn't the last request time out? The server waited two minutes to reply.

**Comment by cstjean** [comment]
On HTTP 0.8.2, that last request yields:

```julia
julia> HTTP.get(url, retry=false, retries=1, readtimeout=15)
ERROR: IOError(EOFError() during request(http://localhost:8082/lock/120))

Stacktrace:
 [1] read_to_buffer(::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::Int64) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/ConnectionPool.jl:207
 [2] readuntil(::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::Function, ::Int64) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/ConnectionPool.jl:226
 [3] readuntil at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/ConnectionPool.jl:224 [inlined]
 [4] readheaders(::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::HTTP.Messages.Response) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/Messages.jl:471
 [5] startread(::HTTP.Streams.Stream{HTTP.Messages.Response,HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}}) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/Streams.jl:153
 [6] macro expansion at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/StreamRequest.jl:65 [inlined]
 [7] macro expansion at ./task.jl:244 [inlined]
 [8] #request#1(::Nothing, ::Nothing, ::Int64, ::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol},NamedTuple{(:retry, :retries),Tuple{Bool,Int64}}}, ::Function, ::Type{HTTP.StreamRequest.StreamLayer}, ::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::HTTP.Messages.Request, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/StreamRequest.jl:54
 [9] #request at ./none:0 [inlined]
 [10] #request#1(::Int64, ::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:iofunction, :retry, :retries),Tuple{Nothing,Bool,Int64}}}, ::Function, ::Type{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}, ::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::HTTP.Messages.Request, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/TimeoutRequest.jl:30
 [11] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:iofunction, :retry, :retries, :readtimeout),Tuple{Nothing,Bool,Int64,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}, ::HTTP.ConnectionPool.Transaction{Sockets.TCPSocket}, ::HTTP.Messages.Request, ::Array{UInt8,1}) at ./none:0
 [12] #request#1(::Nothing, ::Type, ::Int64, ::Base.Iterators.Pairs{Symbol,Any,NTuple{4,Symbol},NamedTuple{(:iofunction, :retry, :retries, :readtimeout),Tuple{Nothing,Bool,Int64,Int64}}}, ::Function, ::Type{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}}, ::HTTP.URIs.URI, ::HTTP.Messages.Request, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/ConnectionRequest.jl:56
 [13] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:iofunction, :retry, :retries, :readtimeout),Tuple{Nothing,Bool,Int64,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}}, ::HTTP.URIs.URI, ::HTTP.Messages.Request, ::Array{UInt8,1}) at ./none:0
 [14] #request#1(::Base.Iterators.Pairs{Symbol,Any,NTuple{4,Symbol},NamedTuple{(:iofunction, :retry, :retries, :readtimeout),Tuple{Nothing,Bool,Int64,Int64}}}, ::Function, ::Type{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}}}, ::HTTP.URIs.URI, ::Vararg{Any,N} where N) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/ExceptionRequest.jl:19
 [15] #request at ./none:0 [inlined]
 [16] #request#1(::VersionNumber, ::String, ::Nothing, ::Nothing, ::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:retry, :retries, :readtimeout),Tuple{Bool,Int64,Int64}}}, ::Function, ::Type{HTTP.MessageRequest.MessageLayer{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/MessageRequest.jl:47
 [17] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:retry, :retries, :readtimeout),Tuple{Bool,Int64,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.MessageRequest.MessageLayer{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at ./none:0
 [18] #request#1(::Int64, ::Bool, ::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:retry, :retries, :readtimeout),Tuple{Bool,Int64,Int64}}}, ::Function, ::Type{HTTP.RedirectRequest.RedirectLayer{HTTP.MessageRequest.MessageLayer{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/RedirectRequest.jl:24
 [19] (::getfield(HTTP, Symbol("#kw##request")))(::NamedTuple{(:retry, :retries, :readtimeout),Tuple{Bool,Int64,Int64}}, ::typeof(HTTP.request), ::Type{HTTP.RedirectRequest.RedirectLayer{HTTP.MessageRequest.MessageLayer{HTTP.ExceptionRequest.ExceptionLayer{HTTP.ConnectionRequest.ConnectionPoolLayer{HTTP.TimeoutRequest.TimeoutLayer{HTTP.StreamRequest.StreamLayer}}}}}}, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at ./none:0
 [20] #request#5(::Base.Iterators.Pairs{Symbol,Integer,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:retry, :retries, :readtimeout),Tuple{Bool,Int64,Int64}}}, ::Function, ::String, ::HTTP.URIs.URI, ::Array{Pair{SubString{String},SubString{String}},1}, ::Array{UInt8,1}) at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/HTTP.jl:300
 [21] #request at ./none:0 [inlined]
 [22] #request#6 at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/HTTP.jl:314 [inlined]
 [23] #request at ./none:0 [inlined] (repeats 2 times)
 [24] #get#13 at /home/cst-jean/.julia/packages/HTTP/wY7Vc/src/HTTP.jl:382 [inlined]
 [25] (::getfield(HTTP, Symbol("#kw##get")))(::NamedTuple{(:retry, :retries, :readtimeout),Tuple{Bool,Int64,Int64}}, ::typeof(HTTP.get), ::String) at ./none:0
 [26] top-level scope at none:0
```

and similarly if I pin HTTP.jl to v0.7.1. Is that the timeout exception? The code emits a warning AFAICT, but I can't see it in the output.

https://github.com/JuliaWeb/HTTP.jl/blob/6531e0b0eb80cc5a9dd2600a1b70dfbf4c6efe6d/src/Servers.jl#L312-L326

**Comment by cstjean** [comment]
It seems to be correctly fixed on 0.8.2. Thank you for all the work on HTTP.jl
